client_plugin/drivers/shared/dockerops/dockerops.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Docker host related operations
+//
+// DockerOps struct holds the docker client based on a certain API version
+// and docker socket. All the operations which require a docker client should
+// be executed through this structure, including docker volume create/remove,
+// docker service start/stop, and docker information retrieve.
+
+package dockerops
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	log "github.com/Sirupsen/logrus"
+	dockerClient "github.com/docker/engine-api/client"
+	dockerTypes "github.com/docker/engine-api/types"
+	"github.com/docker/engine-api/types/swarm"
+)
+
+const (
+	// dockerAPIVersion: docker engine 1.24 and above support this api version
+	dockerAPIVersion = "v1.24"
+	// dockerUSocket: Unix socket on which Docker engine is listening
+	dockerUSocket = "unix:///var/run/docker.sock"
+)
+
+// DockerOps is the interface for docker host related operations
+type DockerOps struct {
+	Dockerd *dockerClient.Client
+}
+
+func NewDockerOps() *DockerOps {
+	var d *DockerOps
+
+	client, err := dockerClient.NewClient(dockerUSocket, dockerAPIVersion, nil, nil)
+	if err != nil {
+		log.WithFields(
+			log.Fields{"error": err},
+		).Error("Failed to create client for Docker ")
+		return nil
+	}
+
+	d = &DockerOps{
+		Dockerd: client,
+	}
+
+	return d
+}
+
+// GetSwarmInfo - returns the node ID and node IP address in swarm cluster
+// also returns if this node is a manager or not
+func (d *DockerOps) GetSwarmInfo() (nodeID string, addr string, isManager bool, err error) {
+	info, err := d.Dockerd.Info(context.Background())
+	if err != nil {
+		return
+	}
+
+	// if node is not in active swarm mode, return error
+	if info.Swarm.LocalNodeState != swarm.LocalNodeStateActive {
```

Comment:
It's ok even we return error here when the swarm state is pending.
The plugin will try to restart again and it will come to here again and again until the state is active :)